jib-core/src/main/java/com/google/cloud/tools/jib/cache/Retry.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright 2019 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.jib.cache;
+
+import com.google.common.base.Preconditions;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Predicate;
+
+/**
+ * A simple class for retrying an action until it succeeds, or has retried too often and failed. By
+ * default the action will be run up to 5 times. The action is deemed successful if it runs to
+ * completion without throwing an exception, and returns true.
+ *
+ * <ul>
+ *   <li>Exceptions are caught and, if deemed {@link #retryOnException(Predicate) retryable} then
+ *       the action will be re-attempted. By default, any exception is considered retryable.
+ *   <li>The retry instance can be configured to {@link #sleep(long, TimeUnit) sleep between
+ *       retries}.
+ *   <li>The maximum retry count {@link #maximumRetries(int) is configurable} (5 times by default).
+ * </ul>
+ *
+ * @param <E> the class of exceptions that may be thrown
+ */
+public class Retry<E extends Exception> {
+
+  /** An Action may throw an exception of type {@code E}. */
+  @FunctionalInterface
+  public interface Action<E extends Exception> {
+    /**
+     * Perform the action.
+     *
+     * @return {@code true} if the action was successful and {@code false} otherwise
+     * @throws E exception thrown during the action
+     */
+    boolean run() throws E;
+  }
+
+  /**
+   * Create a retryable action.
+   *
+   * @param action the action to be run
+   * @return the instance
+   * @param <E> the class of exceptions that may be thrown
+   */
+  public static <E extends Exception> Retry<E> action(Action<E> action) {
+    return new Retry<E>(action);
+  }
+
+  private final Action<E> action;
+  private int maximumRetries = 5;
+  private Predicate<? super E> retryOnException = ignored -> true; // continue to retry
```

Comment:
nit:  I would expect something named `retryOnException` to be true/false. But I have no naming alternative :(

also I'm confused how this is `Predicate<? super E>`, how can the exception thrown be super of E?

Reply:
I already changed it to just be `Predicate<Exception>`.  It's not that the exception will be a super of `E`, but that you can reuse a predicate that takes a `Throwable` (for example).